pallets/gear-builtin/src/lib.rs

```diff
@@ -0,0 +1,368 @@
+        ///
+        ///
+        /// This does computations, therefore we should seek to cache the value at the time of
+        /// a builtin actor registration.
+        pub fn generate_actor_id(builtin_id: BuiltinId) -> ProgramId {
+            hash((SEED, builtin_id).encode().as_slice()).into()
+        }
+    }
+}
+
+impl<T: Config> BuiltinRouterProvider<StoredDispatch, JournalNote, u64> for Pallet<T> {
+    type Router = BuiltinRegistry<T>;
+
+    fn provide() -> BuiltinRegistry<T> {
+        BuiltinRegistry::<T>::new()
+    }
+
+    fn provision_cost() -> u64 {
+        <T as Config>::WeightInfo::provide().ref_time()
+    }
+}
+
+pub struct BuiltinRegistry<T: Config> {
+    pub registry: BTreeMap<ProgramId, BuiltinId>,
+    pub _phantom: sp_std::marker::PhantomData<T>,
+}
+impl<T: Config> BuiltinRegistry<T> {
+    fn new() -> Self {
+        let mut registry = BTreeMap::new();
+        let mut builtin_ids = Vec::with_capacity(16);
+        <T as Config>::BuiltinActor::get_ids(&mut builtin_ids);
+        let builtin_ids_len = builtin_ids.len();
+        for builtin_id in builtin_ids {
+            let actor_id = Pallet::<T>::generate_actor_id(builtin_id);
+            registry.entry(actor_id).or_insert(builtin_id);
+        }
+        assert!(
+            registry.len() == builtin_ids_len,
+            "Duplicate builtin ids detected!"
+        );
+
+        Self {
+            registry,
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<T: Config> BuiltinRouter for BuiltinRegistry<T> {
+    type QueuedDispatch = StoredDispatch;
+    type Output = JournalNote;
+
+    fn lookup(&self, id: &ProgramId) -> Option<BuiltinId> {
+        self.registry.get(id).copied()
+    }
+
+    fn dispatch(&self, dispatch: StoredDispatch, gas_limit: u64) -> Option<Vec<JournalNote>> {
+        let actor_id = dispatch.destination();
+
+        // Try to convert the incoming dispatch into a format that the builtin actor can handle.
+        let builtin_message =
+            <T::Message as TryFrom<StoredDispatch>>::try_from(dispatch.clone()).ok()?;
+
+        // Builtin actors can only execute dispatches of `Handle` kind and only `Handle`
+        // dispatches can end up here (TODO: elaborate).
+        if dispatch.kind() != DispatchKind::Handle {
+            unreachable!("Only handle dispatches can end up here");
+        }
+
+        let mut gas_counter = GasCounter::new(gas_limit);
+
+        let (res, gas_spent) = <T as Config>::BuiltinActor::handle(&builtin_message, gas_limit);
+
+        // We rely on a builtin actor having performed the check for gas limit consistency
+        // and having reported an error if the `gas_limit` was to have been exceeded.
+        // However, to avoid gas tree corruption error, we must not report as spent more gas than
+        // the amount reserved in gas tree (that is, `gas_limit`). Hence (just in case):
+        let gas_spent = gas_spent.min(gas_limit);
```

Comment:
Gas allowance doesn't apply here, I think. Once a builtin actor gets a message it will process it till end since we don't have any rolling gas meter with builtins.

Reply:
This check is exclusively to avoid gas tree corruption error - we can't spend more than had been reserved in the gas tree. And with builtins that's totally possible, because there is no way to break execution half way.